File: lib/src/main/java/com/github/snowdream/toybricks/ToyBricks.java
package com.github.snowdream.toybricks;

import com.github.snowdream.toybricks.annotation.InterfaceLoader;
import com.github.snowdream.toybricks.annotation.impl.DefaultInterfaceLoader;

/**
 * Created by snowdream on 17/2/10.
 */
public class ToyBricks {

    //Default InterfaceLoader
    private static InterfaceLoader sDefaultInterfaceLoader;

    //NewI InterfaceLoader
    private static InterfaceLoader sNewInterfaceLoader;

    public static synchronized <T> T getImplementation(Class<T> clazz) {
        T implementation = null;

        if (sDefaultInterfaceLoader == null){
            sDefaultInterfaceLoader = new DefaultInterfaceLoader();
        }

        if (sNewInterfaceLoader == null){
            sNewInterfaceLoader = sDefaultInterfaceLoader.getImplementation(InterfaceLoader.class);
        }

        if (sNewInterfaceLoader != null){
            implementation = sNewInterfaceLoader.getImplementation(clazz);
        }

        if (implementation == null){
            implementation = sDefaultInterfaceLoader.getImplementation(clazz);
        }

        return implementation;
    }
}

File: lib/src/main/kotlin/com/github/snowdream/toybricks/ToyBricks.kt
package com.github.snowdream.toybricks

import com.github.snowdream.toybricks.annotation.InterfaceLoader
import com.github.snowdream.toybricks.annotation.impl.DefaultInterfaceLoader

/**
 * Created by snowdream on 17/2/10.
 */
class ToyBricks {

        companion object{
            //Default InterfaceLoader
            @JvmField var sDefaultInterfaceLoader: InterfaceLoader = DefaultInterfaceLoader()

            //NewI InterfaceLoader
            @JvmField var sNewInterfaceLoader: InterfaceLoader ?= sDefaultInterfaceLoader.getImplementation(InterfaceLoader::class.java)

            @JvmStatic @Synchronized fun <T> getImplementation(clazz: Class<T>): T {
                var implementation: T?

                implementation = sNewInterfaceLoader?.getImplementation(clazz)

                if (implementation == null) {
                    implementation = sDefaultInterfaceLoader.getImplementation(clazz)
                }

                return implementation!!
            }
        }
}